### ISSUE TITLE
Plugin installation behind forward proxy completely broken in 5.1.1

### ISSUE THREAD

**Issue by tkuther** [elastic/logstash#6528]
Just updated to Logstash version 5.1.1 - installed from elastic YUM repo on CentOS 6, Oracle JDK 1.8.
I do need the output-email plugin.

Reading the docs, the most obvious would be to export HTTP_PROXY and HTTPS_PROXY
```
[root@phsysint08 logstash]# export HTTP_PROXY=http://localhost:3128
[root@phsysint08 logstash]# export HTTPS_PROXY=http://localhost:3128
[root@phsysint08 logstash]# export http_proxy=http://localhost:3128
[root@phsysint08 logstash]# export https_proxy=http://localhost:3128
# curl -s -I https://artifacts.elastic.co
HTTP/1.1 200 Connection established
Connection: close

HTTP/1.1 302 Moved Temporarily
Content-Length: 169
Content-Type: text/html
Date: Thu, 12 Jan 2017 15:07:01 GMT
Location: https://www.elastic.co/downloads/
Server: nginx/1.4.6 (Ubuntu)
x-ngx-hostname: www01
Connection: keep-alive

[root@phsysint08 logstash]# DEBUG=1 bin/logstash-plugin install logstash-output-email
DEBUG: exec /usr/share/logstash/vendor/jruby/bin/jruby --1.9 /usr/share/logstash/lib/pluginmanager/main.rb install logstash-output-email
Looking if package named: logstash-output-email exists at https://artifacts.elastic.co/downloads/logstash-plugins/logstash-output-email/logstash-output-email-5.1.1.zip
Errno::ECONNREFUSED: Connection refused - Connection refused
```
^^ Proxy setting gets ignored

Next, try offline installation. Downloaded the .gem
```
[root@phsysint08 logstash]# DEBUG=1 bin/logstash-plugin install /tmp/logstash-output-email-4.0.3.gem 
DEBUG: exec /usr/share/logstash/vendor/jruby/bin/jruby --1.9 /usr/share/logstash/lib/pluginmanager/main.rb install /tmp/logstash-output-email-4.0.3.gem
Looking if package named: /tmp/logstash-output-email-4.0.3.gem exists at https://artifacts.elastic.co/downloads/logstash-plugins//tmp/logstash-output-email-4.0.3.gem//tmp/logstash-output-email-4.0.3.gem-5.1.1.zip
Errno::ECONNREFUSED: Connection refused - Connection refused
```

Next,trying the unpack & install --local way. Downloaded release .zip from github, then
```
[root@phsysint08 logstash]# ./bin/logstash-plugin unpack --zip /tmp/logstash-output-email-4.0.3.zip 
Unpacking /tmp/logstash-output-email-4.0.3.zip
Unpacked at /usr/share/logstash/vendor/cache
The unpacked plugins can now be installed in local-only mode using bin/logstash-plugin install --local [plugin name]
[root@phsysint08 logstash]# DEBUG=1 bin/logstash-plugin install --local logstash-output-email --version 4.0.3
DEBUG: exec /usr/share/logstash/vendor/jruby/bin/jruby --1.9 /usr/share/logstash/lib/pluginmanager/main.rb install --local logstash-output-email --version 4.0.3
Looking if package named: logstash-output-email exists at https://artifacts.elastic.co/downloads/logstash-plugins/logstash-output-email/logstash-output-email-5.1.1.zip
Errno::ECONNREFUSED: Connection refused - Connection refused
...
```
So it seems to ignore both, --local flag AND the proxy setting. JARS_SKIP='true' makes no difference.

Am I doing something wrong here?

**Comment by tkuther** [comment]
#6529 seems to be the same but on Windows

**Comment by ph** [comment]
@tkuther Would you mind trying on 5.1.2?

**Comment by tkuther** [comment]
@ph Done. Same results for all three methods.

I tried to add some logging in lib/pluginmanager/install.rb at def local_gems? to shed some light on why the local gem method fails, but it seems that doesn't get called at all. I had to comment out that whole PACK strategy begin end block. So the proxy issue seems to be somewhere in lib/pluginmanager/pack_fetch_strategy/repository.rb

After commenting that out and using the local gem:
```Bundle complete! 115 Gemfile dependencies, 204 gems now installed.```

I really love the ELK stack and we're using it since 0.9 and earlier, but that plugin mess all around drives me nuts. Or the forward proxy - forward proxies are so 2016. But "yum install lgstash-output-email" still sounds like icecream.

**Comment by Rgils** [comment]
Same there (linux) in 5.1.2 trying with logstash-input-stomp
Any workaround distinct than copy from another machine who is not behind the proxy?

**Comment by Rgils** [comment]
Solved (mine):
`JRUBY_OPTS='-J-Dhttp.proxyHost=CHANGEME -J-Dhttp.proxyPort=CHANGEME -J-Dhttp.proxyUser=CHANGEME -J-Dhttp.proxyPassword=CHANGEME' DEBUG=1 ./bin/logstash-plugin install logstash-input-stomp`
and remeber to run it with sudo -u logstash (or root if nothing works)

**Comment by JiriKadlcik** [comment]
Just adding to that, if the plugin has any .jar dependencies, you need to set up proxy separately for maven, e.g. in ~/.m2/settings.xml

**Comment by jamesblackburn** [comment]
Seems to be broken on the latest 5.2.1 too

**Comment by tkuther** [comment]
@Rgils thanks! Getting closer. Does proxy auth really work for you?
```
JARS_SKIP='true' JRUBY_OPTS='-J-Dhttps.proxyHost=xxx -J-Dhttps.proxyPort=8080 -J-Dhttp.proxyHost=xxx -J-Dhttp.proxyPort=8080 -J-Dhttp.proxyUser=xxx -J-Dhttp.proxyPassword=xxx' DEBUG=1 ./bin/logstash-plugin install logstash-output-email                                                               DEBUG: exec /usr/share/logstash/vendor/jruby/bin/jruby /usr/share/logstash/lib/pluginmanager/main.rb install logstash-output-email
Looking if package named: logstash-output-email exists at https://artifacts.elastic.co/downloads/logstash-plugins/logstash-output-email/logstash-output-email-5.2.1.zip
Net::HTTPServerException: 407 "Proxy Authentication Required"
              error! at /usr/share/logstash/vendor/jruby/lib/ruby/1.9/net/http.rb:2641
...
```

Just for cross-reference, linking https://github.com/elastic/logstash/issues/2557

**Comment by Rgils** [comment]
It's a bit ditry 
```
https_proxy=http://USER:PASSW@proxy.COMPANY.es:8080 HTTPS_PROXY=http://USER:PASSW@proxy.COMPANY.es:8080 JRUBY_OPTS='-J-Dhttp.proxyHost=proxy.COMPANY.es -J-Dhttp.proxyPort=8080 -J-Dhttp.proxyUser=USER -J-Dhttp.proxyPassword=PASSW' DEBUG=1 ./bin/logstash-plugin install logstash-input-stomp
```
but it works haha:
```
DEBUG: exec /usr/share/logstash/vendor/jruby/bin/jruby /usr/share/logstash/lib/pluginmanager/main.rb install logstash-input-stomp
Looking if package named: logstash-input-stomp exists at https://artifacts.elastic.co/downloads/logstash-plugins/logstash-input-stomp/logstash-input-stomp-5.2.2.zip
Package not found at: https://artifacts.elastic.co/downloads/logstash-plugins/logstash-input-stomp/logstash-input-stomp-5.2.2.zip
Validating logstash-input-stomp
Installing logstash-input-stomp
```

**Comment by ph** [comment]
So far I suspect the following is happening:

- in the install flow we don't actually load any current gems to make sure we don't run into issues like the 'already activated XXX' 
- Even if Logstash has a **jar-dependencies* version lock at **3.1.10**, we actually use the `jar-dependencies` that ship with jruby.

Steps:

1.  Add an actual integration test to make sure we corrrectly fix that issue for good.
2. Force *jar-dependencies* version in the context of installation.


I've completed 1, but step 2 is a bit harder with the nature of the plugin manager.

**Comment by ph** [comment]
OK, my theory doesn't seems to be true, since both version are available at install time. I believe the ordering of the gem load path should take the highest version.

```ruby
[4] pry(#<LogStash::PluginManager::Install>)> Gem::Specification.all.select {|s| s.name =~ /jar-d/ }.collect { |s| [s.name, s.version] }
NOTE: Specification.all called from (pry):4:in `evaluate_ruby'
=> [["jar-dependencies", Gem::Version.new("0.3.10")], ["jar-dependencies", Gem::Version.new("0.2.6")]]
```

**Comment by ph** [comment]
Adding traces in the differents jar-dependencies actually confirms that we load the new one.

**Comment by ph** [comment]
Will be fixed in 5.4 with https://github.com/elastic/logstash/pull/6825

**Comment by discointheair** [comment]
for me works:
`JARS_SKIP='true' JRUBY_OPTS='-J-Dhttps.proxyHost=user:password@proxy.com -J-Dhttps.proxyPort=8080 -J-Dhttp.proxyHost=user:password@proxy.com -J-Dhttp.proxyPort=8080' DEBUG=1 ./bin/logstash-plugin install logstash-output-email`
